Check that the HTTP2 or HTTP3 requests have a content-length header

ogercyprien opened the pull request:
Changes in openresty 1.25 included changes in lua-nginx-module. In particular, [this diff](https://github.com/openresty/lua-nginx-module/commit/86bea01b244938e0ab6eda780906c06fa62c3b5c) seems to introduce a breaking change where `ngx.req.read_body()` triggers an error when an HTTP2 or HTTP3 request is performed with no `Content-Length` header.

The aim of this PR is to fix this behavior.

ogercyprien committed:
fix(request): check that the HTTP2 or HTTP3 requests have a content-length header

This check ensures that the underlying lua-nginx-module will not crash
when trying to fetch the body of a POST request made with HTTP2 or HTTP3
with an unspecified Content-Length header.
See https://github.com/openresty/lua-nginx-module/pull/2237

jleproust commented:
Fix seems good to me, were all the observed instances of this log for now all bogus requests? No way this can happen with a legit request?

ogercyprien commented:
It is possible that this could happen to a legitimate request. However, when analyzing logs, the only entries that fail are either from bots that seem to target invalid content or from scripts which may target valid content but only 5 such instances have been found. You can find the list in [the associated story](https://app.shortcut.com/opendatasoft/story/47780/fix-error-when-an-http2-request-is-performed-without-content-length).
I think that the analysis sample is large enough to consider that there are not legitimate requests with this pattern.

jleproust commented on lib/resty/waf/request.lua:
```diff
@@ -133,6 +133,15 @@ function _M.parse_request_body(waf, request_headers, collections)
+		if ngx.req.http_version() >= 2 and request_headers["content-length"] == nil then
+			logger.warn(waf, "HTTP2/HTTP3 request without a Content-Length header")
+			if waf._mode == "ACTIVE" then
+				ngx.exit(ngx.HTTP_FORBIDDEN)
```
I'd suggest we return a 400 error instead of 403 as this is an RFC-compliance issue, like other errors above.

```suggestion
				ngx.exit(400)
```

ogercyprien replied:
Sure, good idea

jleproust commented on lib/resty/waf/request.lua:
```diff
@@ -133,6 +133,15 @@ function _M.parse_request_body(waf, request_headers, collections)
+		-- https://github.com/openresty/lua-nginx-module/commit/86bea01b244938e0ab6eda780906c06fa62c3b5c
+		if ngx.req.http_version() >= 2 and request_headers["content-length"] == nil then
+			logger.warn(waf, "HTTP2/HTTP3 request without a Content-Length header")
+			if waf._mode == "ACTIVE" then
```
If waf isn't active, we're still crashing? Shouldn't we return an error anyway?

ogercyprien replied:
You're correct, it a 400 status code should be returned regardless of the WAF status to prevent crashes.

jleproust reviewed (commented):
LGTM but still some important questions.

ogercyprien committed:
Apply review comments

Return 400 status code instead of 403 for RFC-compliance issue
Remove check that the WAF is in active mode as this would trigger also
in passive mode
